File: 09/part1.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"strconv"
	"strings"
)

// This code to remain largely unaltered from the state it was used to solve,
// mostly for historical purposes.

func main() {
	f, err := ioutil.ReadFile("input")
	if err != nil {
		log.Panicln(err)
	}

	var lines = strings.Split(string(f), "\n")

	const preambleLen = 25

	var preamble = make([]int, 0, preambleLen)

	for _, line := range lines {
		n, err := strconv.Atoi(line)
		if err != nil {
			continue
		}

		preamble = append(preamble, n)

		if len(preamble) > preambleLen {
			if !isValidSum(preamble, n) {
				fmt.Println(n, "is invalid.")
				break
			}

			// Cut the first out.
			copy(preamble[0:], preamble[1:])
			preamble = preamble[:len(preamble)-1]
		}
	}
}

func isValidSum(preamble []int, n int) bool {
	for _, preambNum1 := range preamble {
		for _, preambNum2 := range preamble {
			if preambNum2+preambNum1 == n {
				return true
			}
		}
	}
	return false
}

File: 09/part1.zig
const std = @import("std");
const stdout = std.io.getStdOut().writer();
var gpallocator = std.heap.GeneralPurposeAllocator(.{}){};
var allocator = &gpallocator.allocator;

// This code was written AFTER the Go iterations.

pub fn main() !void {
    const f = try std.fs.cwd().openFile("input", .{ .read = true });
    const b = try f.reader().readAllAlloc(allocator, 1024000);

    var line_iterator = std.mem.split(b, "\n");
    var numbers_list = std.ArrayList(u64).init(allocator);

    while (true) {
        var line = line_iterator.next() orelse break;
        var lnum = std.fmt.parseInt(u64, line, 10) catch continue;
        try numbers_list.append(lnum);
    }

    var numbers = numbers_list.toOwnedSlice();

    const preamble_len = 25;
    var preamble: [preamble_len]u64 = [1]u64{0} ** preamble_len;

    for (numbers) |n, i| {
        if (i < preamble_len) {
            preamble[i] = n;
            continue;
        }

        if (!is_valid_sum(&preamble, n)) {
            try stdout.print("{} is invalid.\n", .{n});
            break;
        }

        // Shift the array backwards and replace the last element.
        std.mem.copy(u64, preamble[0..], preamble[1..]);
        preamble[preamble_len - 1] = n;
    }
}

fn is_valid_sum(preamble: []u64, sum: u64) bool {
    for (preamble) |n| {
        for (preamble) |m| {
            if (n + m == sum) {
                return true;
            }
        }
    }
    return false;
}
